fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/ResourceReferenceDAO.java

```diff
@@ -756,76 +754,96 @@ public void upsertCommonTokenValues(List<ResourceTokenValueRec> values) {
             // nothing to do
             return;
         }
-        
+
         // Sort the values so we always process in the same order (deadlock protection)
         List<CommonTokenValue> sortedTokenValues = new ArrayList<>(tokenValueSet);
         sortedTokenValues.sort(CommonTokenValue::compareTo);
 
-        // Build a string of parameter values we use in the query to drive the insert statement.
-        // The database needs to know the type when it parses the query, hence the slightly verbose CAST functions:
-        // VALUES ((CAST(? AS VARCHAR(1234)), CAST(? AS INT)), (...)) AS V(common_token_value, parameter_name_id, code_system_id)
-        StringBuilder inList = new StringBuilder(); // for the select query later
-        StringBuilder paramList = new StringBuilder();
-        for (int i=0; i<sortedTokenValues.size(); i++) {
-            if (paramList.length() > 0) {
-                paramList.append(", ");
-            }
-            paramList.append("(CAST(? AS VARCHAR(" + FhirSchemaConstants.MAX_TOKEN_VALUE_BYTES + "))");
-            paramList.append(",CAST(? AS INT))");
+        // Process the data in a window.
+        int idx = 0;
+        int max = sortedTokenValues.size();
 
-            // also build the inList for the select statement later
-            if (inList.length() > 0) {
-                inList.append(",");
-            }
-            inList.append("(?,?)");
+        // The maximum number of query parameters that are available for a particular persistence layer.
+        // There are two '?' parameters declared for each CommonTokenValue.
+        Optional<Integer> maxQuery = translator.maximumQueryParameters();
+        int maxSub;
+        if (maxQuery.isPresent() && (max / 2) > maxQuery.get()) {
```

Comment:
shouldn't that be max * 2, because max is the number of CommonTokenValue elements in the list, and each requires two bind variables.

Reply:
Yes, you are correct